README.md
# CakeResque Documentation website

This `gh-pages` branch contains the CakeResque documentation website sources. 

Since the website is hosted on another server, changes will not be immediate.

You can contribute by updating the documentation, or add new translatioins. Refer to CONTRIBUTING.md for contribution guidelines.

CHANGELOG.md
##Changelog

###**v4.1.1** [2015-01-28]

* [fix] Fix#80: Only use DEBUG_BACKTRACE_IGNORE_ARGS for PHP_VERSION >= 5.3.6

###**v4.1.0** [2014-08-03]

* [new] Fix#57: Add Redis auth
* [fix] Fix#63: CakeResque should now be installed as a dependency to your main app.
* [fix] Fix#62: Optimize `AppShell::perform()`. Update the `perform()` method inside your AppShell class to

		public function perform() {
			$this->initialize();
			return $this->runCommand($this->args[0], $this->args);
		}

* [fix] Stopping a worker should also removed all empty queues associated to it
* Various improvement and bugfixes

> **NOTE**: All CakeResque dependencies are now installed inside your app `Vendor` dir, and not inside `Plugin/CakeResque/vendor` anymore.

###**v4.0.2** [2013-11-06]

* [fix] Uses `App::uses` in CakeResqueBootstrap

###**v4.0.1** [2013-10-29]

* [new] Add setting to set a custom dir for the workers PID files

###**v4.0.0** [2013-10-25]

* [new] Fix #20: Ease personalized plugin configuration [@bar]
* [fix] Fix #33: Don't use `sudo` when the current user is already the target user
* [fix] Default process owner is the one running the webserver, and not the one owning the file
* [fix] Fix #25: Workers now works with exotic directory structure

###**v3.3.8** [2013-08-22]

* [fix] Handle custom cake `app` folder name

###**v3.3.7** [2013-07-07]

* [new] Add `reset` to reset CakeResque internal state
* [removed] Remove `--debug` option, in favor of `--verbose`
* [fix] Some text formatting and other minor fixes

###**v3.3.6** [2013-05-18]

* [new] `restart` now takes into account the workers your stopped and started just before.
* [new] More reliable worker detection on `start` and `startscheduler`
* [fix] Upgrade CakeResqueBootstrap file with Cake2 version
* [new] Add unit tests. Code coverage can be tracked on coveralls.io

> **NOTE**: Workers status are stocked in a different format. Please *stop all your workers* first before updating.


###**v3.3.5** [2013-05-07]

* [fix] Call kill command using an absolute path [bar]

###**v3.3.4** [2013-04-30]

* [fix] Stop worker gracefully by sending it a SIGNAL, instead of killing the process

###**v3.3.3** [2013-04-24]

* [fix] Handle setup with exotic app folder name

###**v3.3.2** [2013-04-17]

* [new] Add `--debug`

###**v3.3.1** [2013-04-14]

* [new] Add `--verbose` option when starting worker. Default log mode downgraded from verbose to normal.

###**v3.3.0** [2013-04-13]

* [new] Add `clear` command to clear queues' jobs

###**v3.2.4** [2013-02-27]

* [fix] Namespace was changed in ResqueScheduler

###**v3.2.3** [2013-02-12]

* [fix] Invalid jobs number for queues stats
* [fix] Add example of using custom user when starting worker

###**v3.2.2** [2013-02-10]

* [fix] 100% test code coverage for existing tests

###**v3.1.1** [2013-02-07]

* [change] More accurate error message when starting workers fail
* [fix] Fix error while loading the Resque Job Creator class

###**v3.1.0** [2013-02-06]

* [new] Add tests & refactor some class to make them more testable
* [new] Use Travis CI
* [change] Update composer to make plugin installable as a CakePHP Plugin

###**v3.0.4** [2013-02-01]

* [new] Display total number of scheduled jobs in stats

###**v3.0.3** [2013-01-30]

* [fix] Load ResqueScheduler library even if scheduler is disabled to prevent class not found

###**v3.0.2** [2013-01-30]

* [fix] Load missing resque library files
* [fix] More accurate detection of workers status
* [change] Reove redundant fiels include

###**v3.0.1** [2013-01-30]

* [change] Rename default log files for consistency

###**v3.0.0** [2013-01-30]

* [new] Add `enqueueAt()` and `enqueueIn()` for scheduled jobs
* [new] Display pending jobs count for each queues in `stats`
* [new] Display message when a queue is not monitored by a worker, in `stats`
* [new] Display error message when resque library files are not found, instead of fatal error

> Scheduled Jobs are disabled by default. Activate them by setting `CakeResque.Scheduler.enabled` to `true` in the bootstrap file.
> Run `composer update` to update your dependencies
> See [upgrade notes]()http://www.kamisama.me/?p=495)

###**v2.2.1** [2012-10-24]

* [new] Add .pot file for i18n. Help for translation are welcomed.

###**v2.2.0** [2012-10-23]

* [fix] Tracking job not working properly
* [new] Display failed job details using `track` when job status is *fail*

> Require php-resque-ex **1.0.14**.
> Update dependencies with `composer update`


###**v2.1.0** [2012-10-16]

* [new] Add `track` command to track a job status

> A new `CakeResque.Job.track` setting has been added to the bootstrap file.
> It's the master value to enable the job tracking status.
> You can also enable/disable tracking on a per-job basis,
> by passing `true`/`false` as fourth argument when queueing job via `CakeResque::enqueue()`.
>
> Job status tracking is disabled by default.
> Job status is only kept for 24 hours.
> *Unknown* will be returned
>
> - when job ID is invalid,
> - when job status is expired,
> - or when job status tracking is disabled.




###**v2.0.0** [2012-10-14]

* [new] Add `pause` command to pause one or all worker
* [new] Add `resume` command to resume one or all paused worker
* [new] Add `cleanup` command to immediately terminate a worker's job
* [change] Add more documentation

###**v1.2.6** [2012-10-08]

* [new] Use your own php-resque library

###**v1.2.5** [2012-10-03]

* [fix] Strict Error warning when checking for existing user

###**v1.2.4** [2012-10-01]

* [new] Log Job ID for DebugKit resque panel

###**v1.2.3** [2012-10-01]

* [fix] Fix composer dependencies

###**v1.2.2** [2012-09-27]

* [new] Enqueuing a job return job id

###**v1.2.1** [2012-09-10]

* [Fix] Log correct method name when processing job

###**v1.2.0** [2012-09-08]

* [new] Add CakeResque proxy to enable jobs logging
> Refactor all your `Resque::enqueue()` call to `CakeResque::enqueue()`, to enable logging.

Config/config.php
<?php
/**
 * CakeResque configuration file
 *
 * Default settings for Resque workers and queues.
 *
 * PHP version 5
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @author        Wan Qi Chen <kami@kamisama.me>
 * @copyright     Copyright 2012, Wan Qi Chen <kami@kamisama.me>
 * @link          http://cakeresque.kamisama.me
 * @package       CakeResque
 * @subpackage	  CakeResque.Config
 * @since         3.4.0
 * @license       MIT License (http://www.opensource.org/licenses/mit-license.php)
 */

/**
 * Configure the default value for Resque
 *
 * ## Mandatory indexes :
 * Redis
 * 		Redis server settings
 * Worker
 * 		Workers default settings
 * Resque
 * 		Default values used to init the php-resque library path
 *
 * ## Optional indexes :
 * Queues
 * 		An array of queues to start with Resque::load()
 * 		Used when you have multiple queues, as you don't need
 * 		to start each queues individually each time you start Resque
 * Env
 * 		Additional environment variables to pass to Resque
 * Log
 * 		Log handler and its arguments, to save the log with Monolog
 *
 *
 * There are many ways to configure the plugin:
 *
 * 1. This file is automagically loaded by the bootstrapping process, when no 'CakeResque'
 * configuration key exists.
 *
 *   CakePlugin::load('CakeResque', array('bootstrap' => true));
 *
 * 2. If a 'CakeResque' configuration key already exists, the default configuration will not be loaded,
 * and the 'CakeResque' key is expected to contain all the values present in the default configuration.
 *
 *   Configure::load('my_cakeresque_config');
 *   CakePlugin::load('CakeResque', array('bootstrap' => true));
 *
 * 3. Another way to configure the plugin is to load it using a custom bootstrap file.
 *
 *   CakePlugin::load('CakeResque', array('bootstrap' => 'my_bootstrap'));
 *
 *   // APP/Plugin/CakeResque/Config/my_bootstrap.php
 *   require_once dirname(__DIR__) . DS . 'Lib' . DS . 'CakeResque.php';
 *   $config = array(); // Custom configuration
 *   CakeResque::init($config);
 *
 * @see CakeResque::init(), CakeResque::loadConfig().
 */
$config['CakeResque'] = array(
	'Redis' => array(
		'host' => 'localhost',		// Redis server hostname
		'port' => 6379,				// Redis server port
		'database' => 0,			// Redis database number
		'namespace' => 'resque',	// Redis keys namespace
		'password' => null			// Redis password
	),

	'Worker' => array(
		'queue' => 'default',		// Name of the default queue
		'interval' => 5,			// Number of second between each poll
		'workers' => 1,				// Number of workers to create
		// 'user' => 'www-data' 	// User running the worker process

		// Path to the log file
		// Can be an
		// - absolute path,
		// - an relative path, that will be relative to
		// 	 app/tmp/logs folder
		// - a simple filename, file will be created inside app/tmp/logs
		'log' => TMP . 'logs' . DS . 'resque-worker-error.log',

		// Log Verbose mode
		// true to log more debugging informations
		// Can also be enabled per worker, by starting with --verbose
		'verbose' => false
	),
	'Job' => array(
		// Whether to track job status
		// Enabling this will allow you to track a job status by its ID
		// Job status are purged after 24 hours
		//
		// You can also define per-job tracking by passing true/false when calling
		// CakeResque::enqueue(), CakeResque::enqueueAt() or CakeResque::enqueueIn()
		'track' => false
	),
	/*
	'Queues' => array(
		array(
			'queue' => 'default',	// Use default values from above for missing interval and count indexes
			'user' => 'www-data'	// If PHP is running as a different user on you webserver
		),
		array(
			'queue' => 'my-second-queue',
			'interval' => 10
		)
	)
	*/
	'Resque' => array(
		// Path to the directory containing the worker PID files
		'tmpdir' => App::pluginPath('CakeResque') . 'tmp' . DS
	),

	// Other usefull environment variable you wish to set
	// Passing a key only will search for its value in the $_SERVER scope
	// eg : array('SERVER_NAME'); => will search for the value in $_SERVER['SERVER_NAME']
	// Passing a key and a value will set the env variable to this value
	// eg : array('ARCH' => 'x64')
	'Env' => array(),

	// Log Handler
	// If saving the logs in a plain text file doesn't suit you
	// you can send them to Mysql, or MongoDB, etc ...
	// In that case, you'll need a handler to manage your logs
	// All logs outputted by resque will go to the handler.
	// The classic log file (above) will still be used, for logging
	// stuff likes php error, or other STDOUT outputted by your job classses
	//
	// php-resque-ex uses Monolog to manage all the logging stuff
	// If you uses the original php-resque library, these settings
	// will be ignored
	//
	// handler
	//		Name of the Handler (the handler classname, without the 'Handler' part)
	// target
	//		Arguments taken by the handler constructor. If the handler required
	//		multiple arguments, separate them with a comma
	//
	// As of now, the following handler are supported:
	//
	// [HANDLER]		[TARGET]
	// Cube 			Cube server address (e.g: udp://127.0.0.1:1180)
	// RotatingFile 	Path to the log file (e.g: /path/to/resque.log)
	// Syslog 			Facility name
	// Socket 			Address (e.g: udp://127.0.0.1:23)
	// MongoDB 			MongoDB server address  (e.g: mongodb://localhost:27017)
	'Log' => array(
		'handler' => 'RotatingFile',
		'target' => TMP . 'logs' . DS . 'resque.log'
	),

	// Scheduler Worker
	// It's the worker handling all the scheduled jobs
	// Only one scheduler worker is permitted to run at one time
	// It can be paused, resumed and stopped like any other workers
	// It can be started only with the `startscheduler` command,
	// or with `load` if Scheduler Worker is enabled.
	//
	// Scheduled jobs requires the php-resque-ex-scheduler library,
	// that should be installed with automatically via the
	// `composer update` or `composer install` command
	//
	// The Scheduler Worker have its own default settings
	//
	// @since 2.3.0
	//
	'Scheduler' => array(
		// Enable or disable delayed job
		'enabled' => false,

		// Path to the log file
		'log' => TMP . 'logs' . DS . 'resque-scheduler-error.log',

		// Optional
		// Will not default to settings defined in the global scope above
		'Env' => array(),

		// Optional
		// Will default to settings defined in the global scope above
		// Only available setting is `interval`
		// The worker will always poll a fixed special queue, and only one worker can run at one time
		'Worker' => array(
			'interval' => 3
		),

		// Optional
		// Will default to settings defined in the global scope above
		'Log' => array(
			'handler' => 'RotatingFile',
			'target' => TMP . 'logs' . DS . 'resque-scheduler.log'
		)
	)
);
